Sources/FoundationInternationalization/Formatting/Date/Date+RelativeFormatStyle.swift

```diff
@@ -151,69 +212,108 @@ extension Date {
+        fileprivate func _largestNonZeroComponent(_ destDate: Date, reference refDate: Date, adjustComponent: ComponentAdjustmentStrategy) -> CalendarComponentAndValue? {
+            // Precision of `Date` is higher than second, which is the smallest supported unit. Round to seconds.
+            let refDate = destDate.addingTimeInterval(refDate.timeIntervalSince(destDate).rounded(increment: 1.0, rule: .toNearestOrAwayFromZero))
 
-            let nanosecondRange = 1.0e+9
-            let dateComponents : DateComponents
-            if let nanosecond = components.value(for: .nanosecond), abs(nanosecond) > Int(0.5 * nanosecondRange), let adjustedDestDate = calendar.date(byAdding: .second, value: nanosecond > 0 ? 1 : -1, to: destDate) {
-                dateComponents = calendar.dateComponents(Set(ICURelativeDateFormatter.sortedAllowedComponents), from: refDate, to: adjustedDestDate)
-            } else {
-                dateComponents = components
+            let components = self.calendar.dateComponents([.year, .second, .nanosecond], from: refDate, to: destDate)
+            let nanosecond = components.nanosecond ?? 0
+
+
+            guard var adjustedDestDate = calendar.date(byAdding: .nanosecond, value: -nanosecond, to: destDate) else {
+                return nil
+            }
+
+            if abs(nanosecond) >= 500000000,
+               let date = calendar.date(byAdding: .second, value: nanosecond > 0 ? 1 : -1, to: adjustedDestDate) {
+                adjustedDestDate = date
+            }
+
+            let allowedComponents = Set(self.sortedAllowedComponents)
+
+            guard let smallest = self.sortedAllowedComponents.last else {
+                return nil
             }
 
+            let dateComponents = self.calendar.dateComponents(allowedComponents, from: refDate, to: adjustedDestDate)
+
             let compAndValue: CalendarComponentAndValue
-            if let largest = dateComponents.nonZeroComponentsAndValue.first {
-                let comp = largest.component
-                if comp == .hour || comp == .minute || comp == .second {
-                    compAndValue = Self._roundedLargestComponentValue(components: dateComponents, for: destDate, calendar: calendar) ?? largest
-                } else {
-                    // 79144218: It's incorrect to simply use `dateComponents` to determine the date difference. For example, two dates that are 23 hour apart may be in the same day or cross two days.
-                    // Adjust the component value using the day/week/month/year boundaries.
-                    compAndValue = Self._alignedComponentValue(component: largest.component, for: destDate, reference: refDate, calendar: calendar) ?? largest
-                }
+            let largest = dateComponents.nonZeroComponentsAndValue.first ?? (smallest, 0)
+
+                        let comp = largest.component
+                        if comp == .hour || comp == .minute || comp == .second {
```

Comment:
NIt: Not sure if it's me locally but the indentation in this block looks off too.

Reply:
not sure what happened there...but it's fixed now